linera-service/src/chain_listener.rs

```diff
@@ -132,92 +140,94 @@ where
             client
         };
         let (_listen_handle, mut local_stream) = client.listen().await?;
-        if let Err(error) = client.lock().await.process_inbox_if_owned().await {
-            warn!(%error, "Failed to process inbox after starting stream.");
-        }
-        while let Some(notification) = local_stream.next().await {
+        let mut timeout = storage.clock().current_time();
+        loop {
+            let sleep = Box::pin(storage.clock().sleep_until(timeout));
+            let notification = match future::select(local_stream.next(), sleep).await {
+                Either::Left((Some(notification), _)) => notification,
+                Either::Left((None, _)) => break,
+                Either::Right(((), _)) => {
+                    match client.lock().await.process_inbox_if_owned().await {
+                        Err(error) => warn!(%error, "Failed to process inbox."),
+                        Ok((_, None)) => timeout = Timestamp::from(u64::MAX),
+                        Ok((_, Some(new_timeout))) => timeout = new_timeout.timestamp,
+                    }
+                    continue;
+                }
+            };
             info!("Received new notification: {:?}", notification);
-            if config.delay_before_ms > 0 {
-                tokio::time::sleep(Duration::from_millis(config.delay_before_ms)).await;
+            Self::maybe_sleep(config.delay_before_ms).await;
+            match &notification.reason {
+                Reason::NewIncomingMessage { .. } => timeout = storage.clock().current_time(),
+                Reason::NewBlock { .. } | Reason::NewRound { .. } => {
+                    if let Err(error) = client.lock().await.update_validators().await {
+                        warn!(
+                            "Failed to update validators about the local chain after \
+                            receiving notification {:?} with error: {:?}",
+                            notification, error
+                        );
+                    }
+                }
             }
+            Self::maybe_sleep(config.delay_after_ms).await;
+            let Reason::NewBlock { hash, .. } = notification.reason else {
+                continue;
```

Comment:
Question: this skips the `update_wallet` call, so I just wanted to check if that's okay for this case.

Reply:
I think it's sufficient to do it only on `NewBlock` notifications:
A new round doesn't change the wallet, and if we `process_inbox_if_owned` and actually did create a new block, we will receive a `NewBlock` notification anyway, so we will update the wallet then.
(That notification is generated by the local node, so we're not relying on the connection or the validators for it.)